src/engine/Bind.cpp

```diff
@@ -0,0 +1,218 @@
+                 &_bind._input);
+             ptr) {
+    size_t columns{_subtree->getVariableColumn(ptr->_var)};
+    result->_resultTypes.push_back(subRes->_resultTypes[columns]);
+    CALL_FIXED_SIZE_2(inwidth, outwidth, Bind::computeRenameBind,
+                      &result->_data, subRes->_data, columns);
+  } else if (auto ptr = std::get_if<GraphPatternOperation::Bind::Constant>(&_bind._input); ptr) {
+    /*
+    if (auto ptrI = std::get_if<string>(&(ptr->_value))) {
+      result->_resultTypes.push_back(ResultTable::ResultType::LOCAL_VOCAB);
+      auto targetVal = result->_localVocab->size();
+      result->_localVocab->push_back(*ptrI);
+      CALL_FIXED_SIZE_2(inwidth, outwidth, Bind::computeConstantBind,
+                        &result->_data, subRes->_data, targetVal);
+                        */
+        result->_resultTypes.push_back(ResultTable::ResultType::VERBATIM);
+        CALL_FIXED_SIZE_2(inwidth, outwidth, Bind::computeConstantBind,
+                          &result->_data, subRes->_data, ptr->_value);
+
+
+
+  } else {
+    AD_THROW(ad_semsearch::Exception::BAD_QUERY,
+             "Currently only renaming and sum binds are implemented");
+  }
+
+  // TODO<joka921> : in case of a renaming operation we are also sorted on
+  // the renamed result, implement this.
+  result->_sortedBy = resultSortedOn();
+
+  LOG(DEBUG) << "Sort result computation done." << endl;
+}
+
+template <int IN_WIDTH, int OUT_WIDTH>
+void Bind::computeSumBind(IdTable* dynRes, const IdTable& inputDyn,
+                          std::array<size_t, 2> columns,
+                          array<ResultTable::ResultType, 2> inputTypes,
+                          const Index& index) {
+  const auto input = inputDyn.asStaticView<IN_WIDTH>();
+  auto res = dynRes->moveToStatic<OUT_WIDTH>();
+
+  const auto inSize = input.size();
+  res.reserve(inSize);
+  const auto inCols = input.cols();
+  // copy the input to the first cols;
+  for (size_t i = 0; i < inSize; ++i) {
+    res.emplace_back();
+    for (size_t j = 0; j < inCols; ++j) {
+      res(i, j) = input(i, j);
+    }
+    float singleRes = .0f;
+    for (size_t colIdx = 0; colIdx < 1; ++colIdx) {
+      if (inputTypes[colIdx] == ResultTable::ResultType::VERBATIM) {
+        singleRes += input(i, columns[colIdx]);
+      } else if (inputTypes[colIdx] == ResultTable::ResultType::FLOAT) {
+        // used to store the id value of the entry interpreted as a float
+        float tmpF;
+        std::memcpy(&tmpF, &input(i, colIdx), sizeof(float));
+        singleRes += tmpF;
+      } else if (inputTypes[colIdx] == ResultTable::ResultType::TEXT ||
+                 inputTypes[colIdx] == ResultTable::ResultType::LOCAL_VOCAB) {
+        singleRes = std::numeric_limits<float>::quiet_NaN();
+      } else {
+        // load the string, parse it as an xsd::int or float
+        // TODO(schnelle): What's the correct way to handle OPTIONAL here
```

Comment:
If the value is unbound the expression should raise a type error (as far as I understand the XPATH specs). Bind is turned into an `Expand` during query evaluation, and `Expand` on a solution set `u` returns `u` itself if the evaluated expr returns an error. In QLever this equals setting the value of the variable defined by bind to `ID_NO_VALUE`

Reply:
You are right, but to handle this correctly, we first need the possibility to have heterogeneous datatypes in the same column, Which is on the roadmap, but quite some work.